src/UpgradeRule/AddNamespaceRule.php

```diff
@@ -77,6 +79,43 @@ public function withRoot($root)
         return $this;
     }
 
+
+    public function beforeUpgrade(CollectionInterface $code, CodeChangeSet $changeset)
```

Comment:
I'm not assuming that all "before upgrade" tasks are done on a per-file basis. It only is in this case.